### PR TITLE
s3token: retry auth request after BadStatusLine error

### PR DESCRIPTION
**Pull request by fvennetier**
Retry keystone request in case of `BadStatusLine` exception (due to undetected remote disconnection).

### CODE REVIEW

File: swift3/s3_token_middleware.py [hunk 1]
```diff
@@ -197,21 +199,34 @@ def _deny_request(self, code):
...
+                self._logger.warning('HTTP request BadStatusLine, retrying')
+                continue
+            except requests.exceptions.RequestException as e:
+                self._logger.info('HTTP connection exception: %s', e)
+                raise self._deny_request('InvalidURI')
+
+            if response.status_code >= 500:
+                self._logger.warning('HTTP request returned %s, retrying',
+                                     response.status_code)
+                continue
```
Comment (AymericDu) [inline swift3/s3_token_middleware.py]: If the last attempt has a status code greater than 500, no exception is raised.

File: swift3/s3_token_middleware.py [hunk 2]
```diff
@@ -197,21 +199,34 @@ def _deny_request(self, code):
...
+                    '%s/v2.0/s3tokens' % self._request_uri, headers=headers,
+                    data=creds_json, verify=self._verify,
+                    timeout=self._timeout)
+            except requests.exceptions.Timeout as e:
+                self._logger.info('HTTP timeout: %s', e)
+                raise self._deny_request('ServiceUnavailable')
+            except httplib.BadStatusLine as e:
+                # See https://github.com/requests/requests/issues/2364
+                if attempt + 1 >= self._max_attempts:
+                    raise
```
Comment (murlock) [inline swift3/s3_token_middleware.py]: Could we raise self._deny_request('ServiceUnavailable') ?

Comment (fvennetier) [reply swift3/s3_token_middleware.py]: Yes, it would be nicer for the upper level.

File: swift3/s3_token_middleware.py (hunk 1 above)
Comment (fvennetier) [reply swift3/s3_token_middleware.py]: I thought the response status would be sufficient. But you are right, we must raise an exception.

**Pull request merged**